site/app/controllers/calendar/CalendarController.php

```diff
@@ -27,8 +31,125 @@ class CalendarController extends AbstractController {
+
+        $gradeables_of_user = GradeableUtils::getAllGradeableListFromUserId($this->core, $user, $courses, $calendar_messages);
+
+        return new WebResponse(CalendarView::class, 'showCalendar', CalendarInfo::loadGradeableCalendarInfo($this->core, $gradeables_of_user, $courses, $calendar_messages));
+    }
+
+    /**
+     * @Route("/courses/{_semester}/{_course}/calendar")
+     * @return WebResponse
+     */
+    public function viewCourseCalendar(): WebResponse {
+        $calendar_messages = [];
+        $user = $this->core->getUser();
+        $courses = $this->core->getQueries()->getCourseForUserId($user->getId());
+        $name = $this->core->getConfig()->getCourse();
+        // Initialize $oneCourse and get the specific course for the course calendar
+        $oneCourse = $courses[0];
+        foreach ($courses as $course) {
+            if ($course->getTitle() == $name) {
+                $oneCourse = [$course];
+                break;
+            }
+        }
+        $gradeables = GradeableUtils::getGradeablesFromUserAndCourse($this->core, $calendar_messages);
+
+        return new WebResponse(CalendarView::class, 'showCalendar', CalendarInfo::loadGradeableCalendarInfo($this->core, $gradeables, $oneCourse, $calendar_messages), true);
+    }
+
+    /**
+     * @Route("/courses/{_semester}/{_course}/calendar/newItem", methods={"POST"})
+     * @AccessControl(role="INSTRUCTOR")
+     */
+    public function createMessage(): RedirectResponse {
+        $type = $_POST['type'];
+        $date = $_POST['date'];
+        $text = $_POST['text'];
+
+        if (strip_tags($text) !== $text) {
+            $this->core->addErrorMessage("HTML cannot be used in this text");
+            return new RedirectResponse($this->core->buildCourseUrl(['calendar']));
+        }
+
+        $calendar_item = new CalendarItem();
+        try {
+            $calendar_item->setStringType($type);
+        }
+        catch (\InvalidArgumentException $e) {
+            $this->core->addErrorMessage("That is not a valid calendar item type");
+            return new RedirectResponse($this->core->buildCourseUrl(['calendar']));
+        }
+        $calendar_item->setDate(new \DateTime($date));
+        try {
+            $calendar_item->setText($text);
+        }
+        catch (\InvalidArgumentException $e) {
+            $this->core->addErrorMessage("Text exceeds 255 characters which is not allowed");
+            return new RedirectResponse($this->core->buildCourseUrl(['calendar']));
+        }
+
+        $this->core->getCourseEntityManager()->persist($calendar_item);
+        $this->core->getCourseEntityManager()->flush();
+
+        $this->core->addSuccessMessage("Calendar item successfully added");
+        return new RedirectResponse($this->core->buildCourseUrl(['calendar']));
+    }
+
+    /**
+     * @Route("/courses/{_semester}/{_course}/calendar/editItem", methods={"POST"})
```

Comment:
This route could also be overhauled.  What about something along the lines of `/courses/{_semester}/{_course}/calendar/items/{id}/edit`?